llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+  // calculation is as follows:
+  // Defs - (Ops - Defs) = 2 * Defs - Ops
+  int StackAdj = (2 * static_cast<int>(MI->getNumExplicitDefs())) -
+                 static_cast<int>(MI->getNumExplicitOperands());
+  StackHeight += StackAdj;
+
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitSWAP(unsigned Depth) {
+  unsigned Opc = EVM::getSWAPOpcode(Depth);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), DebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitDUP(unsigned Depth) {
+  StackHeight += 1;
+  unsigned Opc = EVM::getDUPOpcode(Depth);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), DebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitPOP() {
+  StackHeight -= 1;
+  assert(StackHeight >= 0);
+  auto NewMI =
+      BuildMI(*CurMBB, CurMBB->end(), DebugLoc(), TII->get(EVM::POP_S));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitConstant(const APInt &Val) {
+  StackHeight += 1;
+  unsigned Opc = EVM::getPUSHOpcode(Val);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), DebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)));
+  if (Opc != EVM::PUSH0)
+    NewMI.addCImm(ConstantInt::get(MF.getFunction().getContext(), Val));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitSymbol(const MachineInstr *MI,
+                                                     MCSymbol *Symbol) {
+  unsigned Opc = MI->getOpcode();
+  assert(Opc == EVM::DATASIZE ||
+         Opc == EVM::DATAOFFSET && "Unexpected symbol instruction");
+  StackHeight += 1;
+  // This is codegen-only instruction, that will be converted into PUSH4.
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::getStackOpcode(Opc)))
+                   .addSym(Symbol);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitConstant(uint64_t Val) {
+  emitConstant(APInt(256, Val));
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitLabelReference(
+    const MachineInstr *Call) {
+  assert(Call->getOpcode() == EVM::FCALL && "Unexpected call instruction");
+  StackHeight += 1;
+  auto [It, Inserted] = CallReturnSyms.try_emplace(Call);
+  if (Inserted)
+    It->second = MF.getContext().createTempSymbol("FUNC_RET", true);
+  auto NewMI =
+      BuildMI(*CurMBB, CurMBB->end(), DebugLoc(), TII->get(EVM::PUSH_LABEL))
+          .addSym(It->second);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitFuncCall(const MachineInstr *MI,
+                                                       const GlobalValue *Func,
+                                                       int StackAdj,
```

Comment:
Can't we move `StackAdj`, `WillReturn` computation from the caller to the function itself? They are defined by the function we call, so making them explicit parameters increases fragility.

Reply:
Removed all arguments except MI, and added helper functions to calculate number of inputs and to check whether call returns. Also, added assert to check whether we have enough operands on stack.